### PR TITLE
Jetpack plugin install: Move installer out of thank you modal

### PR DESCRIPTION
**Pull request by sirreal**
Implementing paObgF-fw-p2

## Changes

- Move the component responsible for provisioning plan plugins from the modal to the checklist. This ensures that the component remains mounted when the modal is dismissed or closed allowing for installation to proceed as expected.
- Modify installer to read desired plugin provisioning from query params. #34183 (merged) requests `all` installation after purchasing a plan which matches current behavior.

## Testing

* Checkout flows and My Plan pages should be unaffected for Simple and Atomic sites.
* Set up a Jetpack site on the free plan. The experience should be unchanged.
* Purchase a plan for the Jetpack site. The plan should be correctly provisioned as before (Akismet + VaultPress set up as required).
* Deactivate, remove, or delete plugin keys for Akismet and/or VaultPress on the Jetpack site. Visit `/plans/my-plan/SITE_SLUG?install=INSTALL_VALUE` with different `INSTALL_VALUE`s. Each should perform as expected:
  - `all`
  - `akismet`
  - `vaultpress`
  - `?install=akismet&install=vaultpress` should also work as expected.
  - `?install` or `?install=some-bad-value` should have no negative impact.
* Notice that the modal is no longer required for plugin provisioning to happen.

A [review](https://github.com/Automattic/wp-calypso/pull/34180#pullrequestreview-254896044) discovered that plugin keys may not initially be available, D31235-code is a patch to help testing that case.
1. Apply D31235-code and sandbox `public-api.wordpress.com`
1. Visit `/plans/my-plan/SITE_SLUG?thank-you&install=all`. You'll see repeated network requests for the keys until both are available.
1. Use the changes from D31235-code to make more keys available one at a time or all at once.
1. Plugins should begin provisioning as soon as they have a key if they've been requested.

In this case, progress still may not be reported accurately, but I don't believe that's a regression. I believe if both keys were not available before this patch, installation would never begin. Accurate progress reporting can be improved in a follow-up PR. I'd also suggest removing the progress in the modal and using a primary button to dismiss. It's no longer important for users to remain on the modal with the changes included here.

### CODE REVIEW

**Comment by sirreal** [comment]
Rebased on master with #34183 landed.

**Comment by sirreal** [comment]
There are a few minor lint issues with this PR I'll sort out 🙄

**Comment by sirreal** [comment]
> There are a few minor lint issues with this PR I'll sort out 🙄

Lint errors should be sorted now. Ready for review 🎉

File: client/my-sites/plans/current-plan/jetpack-product-install/index.tsx [hunk 1]
```diff
@@ -233,20 +255,44 @@ export class JetpackProductInstall extends Component {
...
 }
 
+function mapStateToProps( state ) {
+	const siteId = getSelectedSiteId( state );
+	const queryArgs = getCurrentQueryArguments( state );
+
+	const installQuery: string[] =
+		// eslint-disable-next-line no-nested-ternary
+		typeof queryArgs === 'object' && 'install' in queryArgs
+			? Array.isArray( queryArgs.install )
```
Comment (p-jackson) [inline client/my-sites/plans/current-plan/jetpack-product-install/index.tsx]: You could use `_.castArray(queryArgs.install)` to collapse a branch of the "beautiful nested ternary" ;)

**Comment by ockham** [comment]
FWIW, `state.jetpackProductInstall[ siteId ]` is 

image

**Comment by sirreal** [comment]
> Starting with a free plan and upgrading to a paid one (tried Personal and Premium) leaves me stuck at this stage of the modal 🙁

Thanks for catching that. I started looking and I am able to reproduce with your steps with a brand new site.

The issue seems to be that we query the plugin keys once, and they're not available initially. So we don't have plugin keys and don't attempt to install anything, but the progress indicator expects several plugins to be installed.

File: client/my-sites/plans/current-plan/jetpack-product-install/index.tsx [hunk 2]
```diff
@@ -233,20 +255,44 @@ export class JetpackProductInstall extends Component {
...
+				? queryArgs.install
+				: [ queryArgs.install ]
+			: [];
+
+	const requestedInstalls: PluginSlug[] = installQuery.includes( 'all' )
```
Comment (simison) [inline client/my-sites/plans/current-plan/jetpack-product-install/index.tsx]: In Jetpack dash there are some special keywords used for installing plugins called "scan" and "backups":

image

(those "set up" buttons show up in dash when Akismet/Vaultpress are disabled when we already have a plan)

Suppose we don't need to add those terms here and we can handle that in route-redirections?

**Comment by sirreal** [comment]
I believe this is blocked by [this issue](https://github.com/Automattic/wp-calypso/pull/34180#pullrequestreview-254896044) for now.

I won't have time to address that personally in the next week. I haven't put much thought into it, but one idea might be to re-fetch the plugin keys on an interval until the requested plugins are installed.

A limitation of this implementation to keep in mind is that the installer will kick off _once_ for a given set of `?install=` args.

**Comment by sirreal** [comment]
rebased

File: client/my-sites/plans/current-plan/jetpack-product-install/index.tsx (hunk 2 above)
Comment (sirreal) [reply client/my-sites/plans/current-plan/jetpack-product-install/index.tsx]: Yep, I'll handle that in an independent PR

**Comment by sirreal** [comment]
I've rebased and completed a round of work that makes the installer resilient to plugin keys that are initially missing and need to be requested as well as make sure the request fires when new keys come in.

I elected to move to http-data because the existing key mechanism did not allow keys to be re-requested, so in the case that keys were missing when initially requested (the issue @ockham pointed out), there was no way to make the keys re-fetch. The http-data mechanism is designed as a reusable module that will allow a part of the main state bundle to be removed in a subsequent PR.

This is ready for another round of reviews 🙇

File: client/my-sites/plans/current-plan/jetpack-product-install/index.tsx [hunk 3]
```diff
@@ -101,42 +125,72 @@ export class JetpackProductInstall extends Component {
...
-		if ( pluginKeys && pluginKeys.akismet && pluginKeys.vaultpress ) {
-			this.setState( {
-				startedInstallation: true,
-			} );
+		// We need plugin keys to attempt install.
+		if ( ! pluginKeys ) {
+			return;
+		}
+
+		const installerPositionalArguments: ['akismet', 'vaultpress'] = [ 'akismet', 'vaultpress' ];
```
Comment (simison) [inline client/my-sites/plans/current-plan/jetpack-product-install/index.tsx]: Just interesting to note that TypeScript linting/prettier/something doesn't enforce spaces here. 🤷‍♂

File: client/my-sites/plans/current-plan/jetpack-product-install/index.tsx [hunk 4]
```diff
@@ -161,34 +215,57 @@ export class JetpackProductInstall extends Component {
...
 		if ( this.shouldRefetchInstallationStatus() ) {
 			this.retries++;
 		}
 	};
 
+	shouldRequestKeys(): boolean {
+		const { siteId, pluginKeys, requestedInstalls } = this.props;
+
+		if ( ! siteId ) {
+			return false;
```
Comment (simison) [inline client/my-sites/plans/current-plan/jetpack-product-install/index.tsx]: For some reason, I've seen these events fire recently:

https://github.com/Automattic/wp-calypso/blob/ffc35db2991cdca61b9b2539c8b7e25068f0b29d/client/my-sites/plugins/jetpack-plugins-setup/index.jsx#L193-L196

Would this be good place to add tracking for this in the new mechanism?

**Comment by simison** [comment]
Saw this warning in the console:

> index.jsx:94 No success handler or default action supplied to the http action. You probably want one or the other. Path: /jetpack-blogs/#####/keys

**Comment by sirreal** [comment]
I've added some testing instructions with a sandbox to force conditions like the [observed issue.](https://github.com/Automattic/wp-calypso/pull/34180#pullrequestreview-254896044)

**Pull request merged**